lxd/instance/drivers/driver_lxc.go

```diff
@@ -7926,8 +7930,52 @@ func (d *lxc) removeDiskDevices() error {
 	return nil
 }
 
-// Network I/O limits.
-func (d *lxc) setNetworkPriority() error {
+// Adds a netfilter rule that sets skb->priority on all skb's which exit the container.
+func setNetworkPriorityWithNetfilter(d *lxc, networkPriority int) error {
+	// in setNetworkPriorityWithCgroup we set skb->priority on all
+	// skb's that comes from the container cgroup on the enter to
+	// the host interfaces
+	// here we go another way, we set skb->priority on all skb's
+	// which exit the container network namespace once skb exits
+	// from the host peer of the container network device.
+
+	// Check that we at least succeeded to set an entry
+	success := false
+	var err error
+	for devName, devConfig := range d.ExpandedDevices() {
```

Comment:
As this is handled at a per-NIC device level and is set explicitly rather than in relation to other instances, is this something we can move into the per-NIC device drivers rather than handling in the lxc driver?

A bit like we do with the call to `err = d.state.Firewall.InstanceSetupBridgeFilter`?

Reply:
>As this is handled at a per-NIC device level and is set explicitly rather than in relation to other instances, is this something we can move into the per-NIC device drivers rather than handling in the lxc driver?

I thought about that too.

- this logic is not specific to any NIC driver, it's universal
- an old `setNetworkPriorityWithCgroup` is specific to the containers and not applicable to the virtual machines
- theoretically, we can make `setNetworkPriorityWithNetfilter` work for VMs too

this is a good question where I should place this but I decided to keep it in LXC driver because originally this feature was for containers and I decided to avoid more global changes as we can do it in the future if it's needed.

Reply:
I wonder if there is even a need to keep the old cgroup approach now that we can do it via netfilter?

Reply:
> I wonder if there is even a need to keep the old cgroup approach now that we can do it via netfilter?

I think it's a good idea to get rid of it. As cgroup-v1 will die eventually. But it's a good question if we want to do this right now or make this transition smoother.

Reply:
the netfilter approach works with cgroupv1 and cgroupv2 systems right?

Reply:
sure. netfilter does not involve cgroups at all